src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java

```diff
@@ -309,18 +330,25 @@ private void setCachedResultAndResumeToken() {
             cursor = createCursor(sourceConfig, mongoClient);
         }
 
-        try {
-            BsonDocument next = cursor.tryNext();
-            if (next == null && cursor.getServerCursor() == null) {
-                cursor.close();
-                cursor = null;
+        if (cursor != null) {
+            try {
+                BsonDocument next = cursor.tryNext();
+                // The cursor has been closed by the server
+                if (next == null && cursor.getServerCursor() == null) {
+                    invalidatedCursor = true;
```

Comment:
The server closed the cursor - so mark it as invalidated and create a new cursor.